File: shared/src/main/java/com/kidsmovies/shared/models/SyncModels.kt
package com.kidsmovies.shared.models

import com.google.firebase.database.IgnoreExtraProperties

/**
 * Video info synced to Firebase for parent visibility
 */
@IgnoreExtraProperties
data class SyncedVideo(
    val title: String = "",
    val collectionNames: List<String> = emptyList(),
    val isFavourite: Boolean = false,
    val isEnabled: Boolean = true,
    val duration: Long = 0,
    val playbackPosition: Long = 0,
    val lastWatched: Long? = null,
    val thumbnailUrl: String? = null
) {
    // No-arg constructor for Firebase
    constructor() : this("", emptyList(), false, true, 0, 0, null, null)
}

/**
 * Collection info synced to Firebase for parent visibility
 */
@IgnoreExtraProperties
data class SyncedCollection(
    val name: String = "",
    val type: String = "REGULAR", // REGULAR, TV_SHOW, SEASON
    val parentName: String? = null, // For seasons - parent TV show name
    val videoCount: Int = 0,
    val isEnabled: Boolean = true,
    val thumbnailUrl: String? = null
) {
    // No-arg constructor for Firebase
    constructor() : this("", "REGULAR", null, 0, true, null)
}

/**
 * Child device info synced to Firebase
 */
@IgnoreExtraProperties
data class SyncedChildDevice(
    val deviceName: String = "",
    val childUid: String = "",
    val childName: String = "", // Display name for the child
    val lastSeen: Long = 0,
    val appVersion: String = "",
    val isOnline: Boolean = false,
    val currentlyWatching: String? = null, // Video title if currently watching
    val todayWatchTime: Long = 0 // Minutes watched today
) {
    // No-arg constructor for Firebase
    constructor() : this("", "", "", 0, "", false, null, 0)
}

/**
 * Sync request from parent app
 */
@IgnoreExtraProperties
data class SyncRequest(
    val requested: Boolean = false,
    val requestedAt: Long = 0,
    val requestedBy: String = "" // Parent UID
) {
    // No-arg constructor for Firebase
    constructor() : this(false, 0, "")
}

/**
 * Lock command from parent app for individual videos/collections
 */
@IgnoreExtraProperties
data class LockCommand(
    val videoTitle: String? = null,
    val collectionName: String? = null,
    val isLocked: Boolean = false,
    val lockedBy: String = "", // Parent UID
    val lockedAt: Long = 0,
    val warningMinutes: Int = 5, // Warning time before lock takes effect (0 = immediate)
    val allowFinishCurrentVideo: Boolean = false // Allow child to finish current video before lock
) {
    // No-arg constructor for Firebase
    constructor() : this(null, null, false, "", 0, 5, false)
}

/**
 * App-level lock command - locks the entire kids app
 */
@IgnoreExtraProperties
data class AppLockCommand(
    val isLocked: Boolean = false,
    val lockedBy: String = "",
    val lockedAt: Long = 0,
    val unlockAt: Long? = null, // Scheduled unlock time (null = manual unlock only)
    val message: String = "App is locked by parent", // Message shown to child
    val warningMinutes: Int = 5, // Warning before lock takes effect
    val allowFinishCurrentVideo: Boolean = false
) {
    // No-arg constructor for Firebase
    constructor() : this(false, "", 0, null, "App is locked by parent", 5, false)
}

/**
 * Schedule settings for when the app can be used
 */
@IgnoreExtraProperties
data class ScheduleSettings(
    val enabled: Boolean = false,
    val allowedDays: List<Int> = listOf(0, 1, 2, 3, 4, 5, 6), // 0=Sunday, 6=Saturday
    val allowedStartHour: Int = 8, // 8 AM
    val allowedStartMinute: Int = 0,
    val allowedEndHour: Int = 20, // 8 PM
    val allowedEndMinute: Int = 0,
    val timezone: String = "UTC"
) {
    // No-arg constructor for Firebase
    constructor() : this(false, listOf(0, 1, 2, 3, 4, 5, 6), 8, 0, 20, 0, "UTC")
}

/**
 * Daily time limit settings
 */
@IgnoreExtraProperties
data class TimeLimitSettings(
    val enabled: Boolean = false,
    val dailyLimitMinutes: Int = 120, // 2 hours default
    val weekendLimitMinutes: Int = 180, // 3 hours on weekends
    val warningAtMinutesRemaining: Int = 10 // Warn when 10 minutes left
) {
    // No-arg constructor for Firebase
    constructor() : this(false, 120, 180, 10)
}

/**
 * Viewing metrics for a child device
 */
@IgnoreExtraProperties
data class ViewingMetrics(
    val todayWatchTimeMinutes: Long = 0,
    val weekWatchTimeMinutes: Long = 0,
    val totalWatchTimeMinutes: Long = 0,
    val lastWatchDate: String = "", // YYYY-MM-DD format
    val videosWatchedToday: Int = 0,
    val mostWatchedVideo: String? = null,
    val lastVideoWatched: String? = null,
    val lastWatchedAt: Long = 0
) {
    // No-arg constructor for Firebase
    constructor() : this(0, 0, 0, "", 0, null, null, 0)
}

/**
 * Family info stored in Firebase
 */
@IgnoreExtraProperties
data class Family(
    val familyId: String = "",
    val createdAt: Long = 0,
    val createdBy: String = "", // Parent UID who created the family
    val familyName: String = "My Family"
) {
    // No-arg constructor for Firebase
    constructor() : this("", 0, "", "My Family")
}

/**
 * Pairing code for connecting child devices
 */
@IgnoreExtraProperties
data class PairingCode(
    val code: String = "",
    val familyId: String = "",
    val createdAt: Long = 0,
    val expiresAt: Long = 0,
    val createdBy: String = "", // Parent UID
    val used: Boolean = false,
    val usedBy: String? = null // Child device UID that used this code
) {
    // No-arg constructor for Firebase
    constructor() : this("", "", 0, 0, "", false, null)

    fun isExpired(): Boolean = System.currentTimeMillis() > expiresAt
    fun isValid(): Boolean = !used && !isExpired()
}

/**
 * Firebase database paths
 */
object FirebasePaths {
    const val FAMILIES = "families"
    const val PAIRING_CODES = "pairingCodes"
    const val CHILDREN = "children"
    const val VIDEOS = "videos"
    const val COLLECTIONS = "collections"
    const val DEVICE_INFO = "deviceInfo"
    const val SYNC_REQUEST = "syncRequest"
    const val LOCKS = "locks"
    const val APP_LOCK = "appLock"
    const val SCHEDULE = "schedule"
    const val TIME_LIMITS = "timeLimits"
    const val METRICS = "metrics"
    const val SETTINGS = "settings"

    fun familyPath(familyId: String) = "$FAMILIES/$familyId"
    fun childPath(familyId: String, childUid: String) = "$FAMILIES/$familyId/$CHILDREN/$childUid"
    fun childVideosPath(familyId: String, childUid: String) = "${childPath(familyId, childUid)}/$VIDEOS"
    fun childCollectionsPath(familyId: String, childUid: String) = "${childPath(familyId, childUid)}/$COLLECTIONS"
    fun childDeviceInfoPath(familyId: String, childUid: String) = "${childPath(familyId, childUid)}/$DEVICE_INFO"
    fun childLocksPath(familyId: String, childUid: String) = "${childPath(familyId, childUid)}/$LOCKS"
    fun childSyncRequestPath(familyId: String, childUid: String) = "${childPath(familyId, childUid)}/$SYNC_REQUEST"
    fun childAppLockPath(familyId: String, childUid: String) = "${childPath(familyId, childUid)}/$APP_LOCK"
    fun childSchedulePath(familyId: String, childUid: String) = "${childPath(familyId, childUid)}/$SETTINGS/$SCHEDULE"
    fun childTimeLimitsPath(familyId: String, childUid: String) = "${childPath(familyId, childUid)}/$SETTINGS/$TIME_LIMITS"
    fun childMetricsPath(familyId: String, childUid: String) = "${childPath(familyId, childUid)}/$METRICS"
    fun pairingCodePath(code: String) = "$PAIRING_CODES/$code"
}

File: flutter_parent/lib/models/sync_models.dart
/// Data models matching the Firebase Realtime Database structure
/// used by the Android child app (KidsMovies).

class SyncedVideo {
  final String title;
  final List<String> collectionNames;
  final bool isFavourite;
  final bool isEnabled;
  final bool isHidden;
  final int duration;
  final int playbackPosition;
  final int? lastWatched;
  final String? thumbnailUrl;
  final String sourceType; // "local" or "onedrive"
  final String? remoteId;

  SyncedVideo({
    this.title = '',
    this.collectionNames = const [],
    this.isFavourite = false,
    this.isEnabled = true,
    this.isHidden = false,
    this.duration = 0,
    this.playbackPosition = 0,
    this.lastWatched,
    this.thumbnailUrl,
    this.sourceType = 'local',
    this.remoteId,
  });

  factory SyncedVideo.fromMap(Map<dynamic, dynamic> map) {
    return SyncedVideo(
      title: map['title'] as String? ?? '',
      collectionNames: (map['collectionNames'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      isFavourite: map['isFavourite'] as bool? ?? false,
      isEnabled: map['isEnabled'] as bool? ?? true,
      isHidden: map['isHidden'] as bool? ?? false,
      duration: (map['duration'] as num?)?.toInt() ?? 0,
      playbackPosition: (map['playbackPosition'] as num?)?.toInt() ?? 0,
      lastWatched: (map['lastWatched'] as num?)?.toInt(),
      thumbnailUrl: map['thumbnailUrl'] as String?,
      sourceType: map['sourceType'] as String? ?? 'local',
      remoteId: map['remoteId'] as String?,
    );
  }

  String get formattedDuration {
    if (duration <= 0) return '';
    final minutes = duration ~/ 60000;
    final seconds = (duration % 60000) ~/ 1000;
    if (minutes > 0) return '${minutes}m ${seconds}s';
    return '${seconds}s';
  }
}

class SyncedCollection {
  final String name;
  final String type; // "REGULAR", "TV_SHOW", "SEASON", "FRANCHISE"
  final String? parentName;
  final int videoCount;
  final bool isEnabled;
  final bool isHidden;
  final String? thumbnailUrl;

  SyncedCollection({
    this.name = '',
    this.type = 'REGULAR',
    this.parentName,
    this.videoCount = 0,
    this.isEnabled = true,
    this.isHidden = false,
    this.thumbnailUrl,
  });

  factory SyncedCollection.fromMap(Map<dynamic, dynamic> map) {
    return SyncedCollection(
      name: map['name'] as String? ?? '',
      type: map['type'] as String? ?? 'REGULAR',
      parentName: map['parentName'] as String?,
      videoCount: (map['videoCount'] as num?)?.toInt() ?? 0,
      isEnabled: map['isEnabled'] as bool? ?? true,
      isHidden: map['isHidden'] as bool? ?? false,
      thumbnailUrl: map['thumbnailUrl'] as String?,
    );
  }

  bool get isTvShow => type == 'TV_SHOW';
  bool get isSeason => type == 'SEASON';
}

class SyncedChildDevice {
  final String deviceName;
  final String childUid;
  final String childName;
  final int lastSeen;
  final String appVersion;
  final bool isOnline;
  final String? currentlyWatching;
  final int todayWatchTime;

  SyncedChildDevice({
    this.deviceName = '',
    this.childUid = '',
    this.childName = '',
    this.lastSeen = 0,
    this.appVersion = '',
    this.isOnline = false,
    this.currentlyWatching,
    this.todayWatchTime = 0,
  });

  factory SyncedChildDevice.fromMap(Map<dynamic, dynamic> map) {
    return SyncedChildDevice(
      deviceName: map['deviceName'] as String? ?? '',
      childUid: map['childUid'] as String? ?? '',
      childName: map['childName'] as String? ?? '',
      lastSeen: (map['lastSeen'] as num?)?.toInt() ?? 0,
      appVersion: map['appVersion'] as String? ?? '',
      isOnline: map['isOnline'] as bool? ?? false,
      currentlyWatching: map['currentlyWatching'] as String?,
      todayWatchTime: (map['todayWatchTime'] as num?)?.toInt() ?? 0,
    );
  }

  String get displayName =>
      childName.isNotEmpty ? childName : deviceName.isNotEmpty ? deviceName : childUid;

  bool get isRecentlyOnline {
    if (isOnline) return true;
    return DateTime.now().millisecondsSinceEpoch - lastSeen < 5 * 60 * 1000;
  }
}

class LockCommand {
  final String? videoTitle;
  final String? collectionName;
  final bool isLocked;
  final String lockedBy;
  final int lockedAt;
  final int warningMinutes;
  final bool allowFinishCurrentVideo;

  LockCommand({
    this.videoTitle,
    this.collectionName,
    this.isLocked = false,
    this.lockedBy = '',
    this.lockedAt = 0,
    this.warningMinutes = 5,
    this.allowFinishCurrentVideo = false,
  });

  Map<String, dynamic> toMap() => {
        if (videoTitle != null) 'videoTitle': videoTitle,
        if (collectionName != null) 'collectionName': collectionName,
        'isLocked': isLocked,
        'lockedBy': lockedBy,
        'lockedAt': lockedAt,
        'warningMinutes': warningMinutes,
        'allowFinishCurrentVideo': allowFinishCurrentVideo,
      };
}

class AppLockCommand {
  final bool isLocked;
  final String lockedBy;
  final int lockedAt;
  final int? unlockAt;
  final String message;
  final int warningMinutes;
  final bool allowFinishCurrentVideo;

  AppLockCommand({
    this.isLocked = false,
    this.lockedBy = '',
    this.lockedAt = 0,
    this.unlockAt,
    this.message = '',
    this.warningMinutes = 5,
    this.allowFinishCurrentVideo = false,
  });

  factory AppLockCommand.fromMap(Map<dynamic, dynamic> map) {
    return AppLockCommand(
      isLocked: map['isLocked'] as bool? ?? false,
      lockedBy: map['lockedBy'] as String? ?? '',
      lockedAt: (map['lockedAt'] as num?)?.toInt() ?? 0,
      unlockAt: (map['unlockAt'] as num?)?.toInt(),
      message: map['message'] as String? ?? '',
      warningMinutes: (map['warningMinutes'] as num?)?.toInt() ?? 5,
      allowFinishCurrentVideo:
          map['allowFinishCurrentVideo'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toMap() => {
        'isLocked': isLocked,
        'lockedBy': lockedBy,
        'lockedAt': lockedAt,
        if (unlockAt != null) 'unlockAt': unlockAt,
        'message': message,
        'warningMinutes': warningMinutes,
        'allowFinishCurrentVideo': allowFinishCurrentVideo,
      };
}

class ScheduleSettings {
  final bool enabled;
  final List<int> allowedDays;
  final int allowedStartHour;
  final int allowedStartMinute;
  final int allowedEndHour;
  final int allowedEndMinute;
  final String timezone;

  ScheduleSettings({
    this.enabled = false,
    this.allowedDays = const [0, 1, 2, 3, 4, 5, 6],
    this.allowedStartHour = 8,
    this.allowedStartMinute = 0,
    this.allowedEndHour = 20,
    this.allowedEndMinute = 0,
    this.timezone = 'UTC',
  });

  factory ScheduleSettings.fromMap(Map<dynamic, dynamic> map) {
    return ScheduleSettings(
      enabled: map['enabled'] as bool? ?? false,
      allowedDays: (map['allowedDays'] as List<dynamic>?)
              ?.map((e) => (e as num).toInt())
              .toList() ??
          [0, 1, 2, 3, 4, 5, 6],
      allowedStartHour: (map['allowedStartHour'] as num?)?.toInt() ?? 8,
      allowedStartMinute: (map['allowedStartMinute'] as num?)?.toInt() ?? 0,
      allowedEndHour: (map['allowedEndHour'] as num?)?.toInt() ?? 20,
      allowedEndMinute: (map['allowedEndMinute'] as num?)?.toInt() ?? 0,
      timezone: map['timezone'] as String? ?? 'UTC',
    );
  }

  Map<String, dynamic> toMap() => {
        'enabled': enabled,
        'allowedDays': allowedDays,
        'allowedStartHour': allowedStartHour,
        'allowedStartMinute': allowedStartMinute,
        'allowedEndHour': allowedEndHour,
        'allowedEndMinute': allowedEndMinute,
        'timezone': timezone,
      };
}

class TimeLimitSettings {
  final bool enabled;
  final int dailyLimitMinutes;
  final int weekendLimitMinutes;
  final int warningAtMinutesRemaining;

  TimeLimitSettings({
    this.enabled = false,
    this.dailyLimitMinutes = 120,
    this.weekendLimitMinutes = 180,
    this.warningAtMinutesRemaining = 10,
  });

  factory TimeLimitSettings.fromMap(Map<dynamic, dynamic> map) {
    return TimeLimitSettings(
      enabled: map['enabled'] as bool? ?? false,
      dailyLimitMinutes: (map['dailyLimitMinutes'] as num?)?.toInt() ?? 120,
      weekendLimitMinutes:
          (map['weekendLimitMinutes'] as num?)?.toInt() ?? 180,
      warningAtMinutesRemaining:
          (map['warningAtMinutesRemaining'] as num?)?.toInt() ?? 10,
    );
  }

  Map<String, dynamic> toMap() => {
        'enabled': enabled,
        'dailyLimitMinutes': dailyLimitMinutes,
        'weekendLimitMinutes': weekendLimitMinutes,
        'warningAtMinutesRemaining': warningAtMinutesRemaining,
      };
}

class ViewingMetrics {
  final int todayWatchTimeMinutes;
  final int weekWatchTimeMinutes;
  final int totalWatchTimeMinutes;
  final String? lastWatchDate;
  final int videosWatchedToday;
  final String? mostWatchedVideo;
  final String? lastVideoWatched;
  final int? lastWatchedAt;

  ViewingMetrics({
    this.todayWatchTimeMinutes = 0,
    this.weekWatchTimeMinutes = 0,
    this.totalWatchTimeMinutes = 0,
    this.lastWatchDate,
    this.videosWatchedToday = 0,
    this.mostWatchedVideo,
    this.lastVideoWatched,
    this.lastWatchedAt,
  });

  factory ViewingMetrics.fromMap(Map<dynamic, dynamic> map) {
    return ViewingMetrics(
      todayWatchTimeMinutes:
          (map['todayWatchTimeMinutes'] as num?)?.toInt() ?? 0,
      weekWatchTimeMinutes:
          (map['weekWatchTimeMinutes'] as num?)?.toInt() ?? 0,
      totalWatchTimeMinutes:
          (map['totalWatchTimeMinutes'] as num?)?.toInt() ?? 0,
      lastWatchDate: map['lastWatchDate'] as String?,
      videosWatchedToday:
          (map['videosWatchedToday'] as num?)?.toInt() ?? 0,
      mostWatchedVideo: map['mostWatchedVideo'] as String?,
      lastVideoWatched: map['lastVideoWatched'] as String?,
      lastWatchedAt: (map['lastWatchedAt'] as num?)?.toInt(),
    );
  }

  String formatWatchTime(int minutes) {
    if (minutes >= 60) {
      return '${minutes ~/ 60}h ${minutes % 60}m';
    }
    return '${minutes}m';
  }
}

class Family {
  final String familyId;
  final int createdAt;
  final String createdBy;
  final String familyName;

  Family({
    required this.familyId,
    this.createdAt = 0,
    this.createdBy = '',
    this.familyName = '',
  });

  factory Family.fromMap(String id, Map<dynamic, dynamic> map) {
    return Family(
      familyId: id,
      createdAt: (map['createdAt'] as num?)?.toInt() ?? 0,
      createdBy: map['createdBy'] as String? ?? '',
      familyName: map['familyName'] as String? ?? '',
    );
  }
}

class PairingCode {
  final String code;
  final String familyId;
  final int createdAt;
  final int expiresAt;
  final String createdBy;
  final bool used;
  final String? usedBy;

  PairingCode({
    required this.code,
    required this.familyId,
    this.createdAt = 0,
    this.expiresAt = 0,
    this.createdBy = '',
    this.used = false,
    this.usedBy,
  });

  factory PairingCode.fromMap(String code, Map<dynamic, dynamic> map) {
    return PairingCode(
      code: code,
      familyId: map['familyId'] as String? ?? '',
      createdAt: (map['createdAt'] as num?)?.toInt() ?? 0,
      expiresAt: (map['expiresAt'] as num?)?.toInt() ?? 0,
      createdBy: map['createdBy'] as String? ?? '',
      used: map['used'] as bool? ?? false,
      usedBy: map['usedBy'] as String?,
    );
  }

  bool get isExpired =>
      DateTime.now().millisecondsSinceEpoch > expiresAt || used;

  Map<String, dynamic> toMap() => {
        'code': code,
        'familyId': familyId,
        'createdAt': createdAt,
        'expiresAt': expiresAt,
        'createdBy': createdBy,
        'used': used,
        if (usedBy != null) 'usedBy': usedBy,
      };
}

/// Firebase path constants
class FirebasePaths {
  static String familyPath(String familyId) => 'families/$familyId';
  static String childPath(String familyId, String childUid) =>
      'families/$familyId/children/$childUid';
  static String childVideosPath(String familyId, String childUid) =>
      'families/$familyId/children/$childUid/videos';
  static String childCollectionsPath(String familyId, String childUid) =>
      'families/$familyId/children/$childUid/collections';
  static String childDeviceInfoPath(String familyId, String childUid) =>
      'families/$familyId/children/$childUid/deviceInfo';
  static String childLocksPath(String familyId, String childUid) =>
      'families/$familyId/children/$childUid/locks';
  static String childAppLockPath(String familyId, String childUid) =>
      'families/$familyId/children/$childUid/appLock';
  static String childSchedulePath(String familyId, String childUid) =>
      'families/$familyId/children/$childUid/settings/schedule';
  static String childTimeLimitsPath(String familyId, String childUid) =>
      'families/$familyId/children/$childUid/settings/timeLimits';
  static String childMetricsPath(String familyId, String childUid) =>
      'families/$familyId/children/$childUid/metrics';
  static String childSyncRequestPath(String familyId, String childUid) =>
      'families/$familyId/children/$childUid/syncRequest';
  static String oneDriveConfigPath(String familyId) =>
      'families/$familyId/oneDriveConfig';
  static String pairingCodePath(String code) => 'pairingCodes/$code';
}
